signature/jws/envelope.go

```diff
@@ -0,0 +1,215 @@
+	var payload jwt.MapClaims
+	if err = json.Unmarshal(req.Payload.Content, &payload); err != nil {
+		return nil, &signature.InvalidSignRequestError{
+			Msg: fmt.Sprintf("payload format error: %v", err.Error())}
+	}
+
+	// JWT sign and get certificate chain
+	compact, certs, err := sign(payload, signedAttrs, method)
+	if err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// generate envelope
+	env, err := generateJWS(compact, req, certs)
+	if err != nil {
+		return nil, err
+	}
+
+	encoded, err := json.Marshal(env)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	e.internalEnvelope = env
+	return encoded, nil
+}
+
+// Verify verifies the envelope and returns its enclosed payload and signer info.
+func (e *envelope) Verify() (*signature.EnvelopeContent, error) {
+	if e.internalEnvelope == nil {
+		return nil, &signature.SignatureEnvelopeNotFoundError{}
+	}
+
+	if len(e.internalEnvelope.Header.CertChain) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "certificate chain is not set"}
+	}
+
+	cert, err := x509.ParseCertificate(e.internalEnvelope.Header.CertChain[0])
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: "malformed leaf certificate"}
+	}
+
+	// verify JWT
+	compact := compactJWS(e.internalEnvelope)
+	if err = verifyJWT(compact, cert.PublicKey); err != nil {
+		return nil, err
+	}
+
+	return e.Content()
+}
+
+// Content returns the payload and signer information of the envelope.
+// Content is trusted only after the successful call to `Verify()`.
+func (e *envelope) Content() (*signature.EnvelopeContent, error) {
+	// extract payload
+	payload, err := e.payload()
+	if err != nil {
+		return nil, err
+	}
+
+	// extract signer info
+	signerInfo, err := e.signerInfo()
+	if err != nil {
+		return nil, err
+	}
+	return &signature.EnvelopeContent{
+		SignerInfo: *signerInfo,
+		Payload:    *payload,
+	}, nil
+}
+
+// payload returns the payload of JWS envelope.
+func (e *envelope) payload() (*signature.Payload, error) {
+	if e.internalEnvelope == nil {
+		return nil, &signature.SignatureEnvelopeNotFoundError{}
+	}
+	// parse protected header to get payload context type
+	protected, err := parseProtectedHeaders(e.internalEnvelope.Protected)
+	if err != nil {
+		return nil, err
+	}
+
+	payload, err := base64.RawURLEncoding.DecodeString(e.internalEnvelope.Payload)
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{
+			Msg: fmt.Sprintf("payload error: %v", err)}
+	}
+
+	return &signature.Payload{
+		Content:     payload,
+		ContentType: protected.ContentType,
+	}, nil
+}
+
+// signerInfo returns the SignerInfo of JWS envelope.
+func (e *envelope) signerInfo() (*signature.SignerInfo, error) {
+	if e.internalEnvelope == nil {
+		return nil, &signature.SignatureEnvelopeNotFoundError{}
+	}
+	var signerInfo signature.SignerInfo
+
+	// parse protected headers
+	protected, err := parseProtectedHeaders(e.internalEnvelope.Protected)
```

Comment:
protected, err := parseProtectedHeaders(e.internalEnvelope.Protected) is called both in payload() and signerInfo(), so when we call e.Content(), we are actually parsing the Protected Headers twice.

Reply:
Thanks! It's redundant.